src/porepy/numerics/ad/functions.py

```diff
@@ -141,6 +142,35 @@ def l2_norm(dim: int, var: pp.ad.AdArray) -> pp.ad.AdArray:
     return pp.ad.AdArray(vals, jac)
 
 
+def safe_power(power: float, zero_val: float, var: AdArray) -> AdArray:
+    """Safe (negative) power of an AdArray.
+
+    The power is performed only for nonzeros in the variable, whereas zeros
+    are left as zeros. This is useful for avoiding division by zero in the
+    Jacobian.
+
+    Parameters:
+        numerator: AdArray representing the numerator.
+        denominator: AdArray representing the denominator.
+
+    Returns:
+        AdArray representing the safe division.
+
+    """
+    if isinstance(var, np.ndarray):
+        _val = var
+    else:
+        _val = var.val
+
+    nonzero_inds = np.abs(_val) > 1e-15
```

Comment:
Is it possible to give a rational behind this number, besides 'what else can it be'? Is it meaningful to make this a user controlled parameter with a default?

Reply:
I'm not sure the default will work, since the AD functions are constructed using partial. But if you want me to, I can certainly set it as a mandatory third argument.

Reply:
Good point. My fear is that if this manifests as a bug at some point, it will be extremely hard to track down. Perhaps a mandatory argument is the safest option, so I think I lean in that direction.